### PR TITLE
Added type definitions

### PR DESCRIPTION
**Pull request by eterna2**
**Changes**
- Added `redux` as dev dependencies for their typing.
- Added type definitions for each of the `pwa-helpers` modules.
- Added `tests` folder with a sample typescript file for all the `pwa-helpers`.

### CODE REVIEW

**Comment by Dabolus** [comment]
Not sure if things have changed since then, but according to the comments in [this PR](https://github.com/Polymer/pwa-helpers/pull/24), it seems like the team would like to keep the typings on [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped) (i.e. `@types`) instead of inside this repo

**Comment by keanulee** [comment]
@Dabolus We are revisiting that decision due to our goal of making a TypeScript template of pwa-starter-kit (which requires its deps to have types, so we would have to maintain the types anyways). I made some yet-to-be-reviewed progress in the [typescript branch](https://github.com/Polymer/pwa-helpers/tree/typescript) (source in TS), but I'll take a look at this PR as well.

File: connect-mixin.d.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    _stateChanged(state) {
+      this.count = state.data.count;
+    }
+  }
+*/
+export declare function connect<S, A extends Action = AnyAction>(
+  store: Store<S, A> | ILazyReducerStore<S, A>
+): <B>(baseElement: B) => B & IElementWithReduxStore<S>;
+export declare function connect<S, A extends Action = AnyAction>(
+  store: ILazyReducerStore<S, A>
```
Comment (keanulee) [inline connect-mixin.d.ts]: Is this second declaration redundant with the first one (since argument `store` is `Store<S, A> | ILazyReducerStore<S, A>`)?

**Comment by Dabolus** [comment]
@keanulee happy to hear that! Having the typings maintained on DefinitelyTyped would slow up **a lot** the work of TypeScript developers, as new updates on `pwa-helpers` would require some days to be published on `@types/pwa-helpers`

File: lazy-reducer-enhancer.d.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+  Then, in your page/element, you can lazy load a specific reducer with:
+  store.addReducers({
+    someReducer
+  });
+*/
+export declare function lazyReducerEnhancer<
+  S,
+  A extends Action = AnyAction,
+  Ext = {},
+  StateExt = {}
```
Comment (keanulee) [inline lazy-reducer-enhancer.d.ts]: I had trouble getting the typings to work correctly when using _both_ redux-thunk _and_ lazy-reducer enhancers (i.e. `compose(lazyReducerEnhancer(combineReducers), applyMiddleware(thunk))`). I [ended up removing lazy-reducer-enhancer entirely](https://github.com/Polymer/pwa-starter-kit/commit/869f01a990ec357ee1d65ec82125cff5abfc3588) and just exported an `addReducers` function from store.js which calls `store.replaceReducer()` itself. Looking at [your fork](https://github.com/e2forks/pwa-starter-kit/tree/template-typescript), seems like you had to do a similar workaround, right?

**Comment by Dabolus** [comment]
Sooo some brainstorming:

## Pros of having the source code as TS

**Robustness**
Having a statically typed language helps spotting and avoiding some errors during development and compilation instead of during runtime, making the code base much more robust

**Consistency**
Most of the other modern Polymer projects are written in TypeScript (lit-html, LitElement, the entire Polymer tools repo and PRPL server). [Even Material Web Components are migrating to TypeScript](https://github.com/material-components/material-components-web-components/pull/109). Having a TS code base would make this repo consistent with others

## Cons

**Overkill?**
TypeScript might be a bit overkill for such a simple code base. However, I don't think this could be considered a real con

**Additional step**
The code needs a build step to be converted into JavaScript. However, this might even be considered a pro, since with a build step we would be able to serve not only an ESM version of the helpers, but also a browser-ready UMD version and a CommonJS version. Also, if we let Travis do the work we don't even have to worry about this extra step

--------------

In conclusion, I see only pros in having the source code as TS. The only con is that we need to configure the build steps properly, but it shouldn't give too much hassle if we use something like Rollup or Parcel

File: lazy-reducer-enhancer.d.ts (hunk 2 above)
Comment (eterna2) [reply lazy-reducer-enhancer.d.ts]: yup. I have the same problem. 

My solution is kinda similar. The only difference is that i wrapped the `addReducers` function to return a `store` that is typed cast to the relevant `State` and `Action` types based on the `reducerMapObject` argument - so that I do not need to have a "global" state.

The issue is with the type definition for `compose` - it doesn't merge `Ext` and `StateExt`, and the last func has no awareness of the prior functions. I could possibly write another set of `compose` declaration where the arguments are `StoreEnhancers` instead of generic functions. This will probably solve that.

```
export function compose(
  f1: StoreEnhancer<Ext0, StateExt0>, f2: StoreEnhancer<Ext1, StateExt1>
): StoreEnhancer<Ext0 & Ext1, StateExt0 & StateExt1>;
```
But it is not really scalable as there are so many combinations.

**Comment by eterna2** [comment]
I totally agree with @Dabolus on his analysis on having typescript source.

File: lazy-reducer-enhancer.d.ts (hunk 2 above)
Comment (keanulee) [reply lazy-reducer-enhancer.d.ts]: This could actually be fine for pwa-starter-kit - we could require users to specify this typing in their store.js based on however many args they use: https://github.com/Polymer/pwa-starter-kit/commit/7a4b58f15d0df315543bfaae1bc80263339b1c2c#diff-cfb8e44b6757fd740446e9e4486b73c1R47
